thumbhash.cpp

```diff
@@ -0,0 +1,292 @@
+                                                        static_cast<float>(std::max(w, h)))),
+                         static_cast<size_t>(1));
+
+    FloatArray l, p, q, a;
+    l.reserve(w * h);
+    p.reserve(w * h);
+    q.reserve(w * h);
+    a.reserve(w * h);
+
+    if (frame->type() == CV_8UC4) {
+        // 4 channels (BGRA)
+        for (int i = 0; i < h; ++i) {
+            for (int j = 0; j < w; ++j) {
+                size_t orig_i = static_cast<size_t>(i * row_ratio);
+                size_t orig_j = static_cast<size_t>(j * col_ratio);
+                const cv::Vec4b& pixel = frame->at<cv::Vec4b>(orig_i, orig_j);
+                float alpha = static_cast<float>(pixel[3]) / 255.0f;                        // A
+                float b =
+                  avg_b * (1.0f - alpha) + (alpha / 255.0f) * static_cast<float>(pixel[0]); // B
+                float g =
+                  avg_g * (1.0f - alpha) + (alpha / 255.0f) * static_cast<float>(pixel[1]); // G
+                float r =
+                  avg_r * (1.0f - alpha) + (alpha / 255.0f) * static_cast<float>(pixel[2]); // R
+                l.push_back((r + g + b) / 3.0f);
+                p.push_back((r + g) / 2.0f - b);
+                q.push_back(r - g);
+                a.push_back(alpha);
+            }
+        }
+    }
+    else if (frame->type() == CV_8UC3) {
+        // 3 channels (BGR)
+        for (int i = 0; i < h; ++i) {
+            for (int j = 0; j < w; ++j) {
+                size_t orig_i = static_cast<size_t>(i * row_ratio);
+                size_t orig_j = static_cast<size_t>(j * col_ratio);
+                const cv::Vec3b& pixel = frame->at<cv::Vec3b>(orig_i, orig_j);
+                float b = (1.0f / 255.0f) * static_cast<float>(pixel[0]); // B
+                float g = (1.0f / 255.0f) * static_cast<float>(pixel[1]); // G
+                float r = (1.0f / 255.0f) * static_cast<float>(pixel[2]); // R
+                l.push_back((r + g + b) / 3.0f);
+                p.push_back((r + g) / 2.0f - b);
+                q.push_back(r - g);
+                a.push_back(1.0f);
+            }
+        }
+    }
+    else if (frame->type() == CV_8U) {
+        for (int i = 0; i < h; ++i) {
+            for (int j = 0; j < w; ++j) {
+                size_t orig_i = static_cast<size_t>(i * row_ratio);
+                size_t orig_j = static_cast<size_t>(j * col_ratio);
+                uchar pixel = frame->at<uchar>(orig_i, orig_j);
+                float l_val = static_cast<float>(pixel) / 255.0f;
+                l.push_back(l_val);
+                p.push_back(0.0f);
+                q.push_back(0.0f);
+                a.push_back(1.0f);
+            }
+        }
+    }
+    else {
+        // Unsupported format
+        return -1;
+    }
+
+    float l_dc, l_scale, p_dc, p_scale, q_dc, q_scale, a_dc, a_scale;
+    FloatArray l_ac, p_ac, q_ac, a_ac;
+    std::tie(l_dc, l_ac, l_scale) = encode_channel(
```

Comment:
super nit but if we use modern enough c++ compiler (C++17) then this can be `auto [l_dc, l_ac, l_scale] = encode_channel(...)` to use the new "structured bindings" feature in C++17

Reply:
currently on c++11... hesitant to change it for this.

https://github.com/discord/lilliput/blob/b084235d2abb83198a68daea2c2153921aaff7e1/opencv.go#L6